Static linking and "DllMain"

Back in 2016, when I tried to fix the static linking issue of OpenBLAS (#813), I did not make any changes to the DllMain function, because as a newcomer, I was not sure how DllMain was being used in the project.

Recently, when I tried to use OpenBLAS together with the [VC-LTL](https://github.com/Chuyu-Team/VC-LTL5) library to make my program run on Win7/XP, I encountered this: 
```
error LNK2005: DllMain already defined in openblas.lib(memory.obj)
```

I've noticed that similar issues have appeared many times both on the web and in this repo. As I read this part of the code again, an idea I had back then came to me once more: when building as static library, should we give `DllMain` another name, such as `StaticLibMain` (as we all know, it is not preferred to have a `DllMain` symbol in a *static* library)? In this way, it wouldn't affect the TLS callback logic at all, and the linking problem can be avoided.

--- a/driver/others/memory.c
+++ b/driver/others/memory.c
@@ -1601,7 +1601,7 @@ void DESTRUCTOR gotoblas_quit(void) {
 }
 
 #if defined(_MSC_VER) && !defined(__clang__)
-BOOL APIENTRY DllMain(HMODULE hModule, DWORD  ul_reason_for_call, LPVOID lpReserved)
+BOOL APIENTRY OpenBLASDllMain(HMODULE hModule, DWORD  ul_reason_for_call, LPVOID lpReserved)
 {
   switch (ul_reason_for_call)
   {
@@ -1650,10 +1650,10 @@ static int on_process_term(void)
 #endif
 
 #ifdef _WIN64
-static const PIMAGE_TLS_CALLBACK dll_callback(HINSTANCE h, DWORD ul_reason_for_call, PVOID pv) = DllMain;
+static const PIMAGE_TLS_CALLBACK dll_callback(HINSTANCE h, DWORD ul_reason_for_call, PVOID pv) = OpenBLASDllMain;
 #pragma const_seg()
 #else
-static void (APIENTRY *dll_callback)(HINSTANCE h, DWORD ul_reason_for_call, PVOID pv) = DllMain;
+static void (APIENTRY *dll_callback)(HINSTANCE h, DWORD ul_reason_for_call, PVOID pv) = OpenBLASDllMain;
 #pragma data_seg()
 #endif
 
@@ -3473,7 +3473,7 @@ void DESTRUCTOR gotoblas_quit(void) {
 }
 
 #if defined(_MSC_VER) && !defined(__clang__)
-BOOL APIENTRY DllMain(HMODULE hModule, DWORD  ul_reason_for_call, LPVOID lpReserved)
+BOOL APIENTRY OpenBLASDllMain(HMODULE hModule, DWORD  ul_reason_for_call, LPVOID lpReserved)
 {
   switch (ul_reason_for_call)
   {
@@ -3517,7 +3517,7 @@ static int on_process_term(void)
 #else
 #pragma data_seg(".CRT$XLB")
 #endif
-static void (APIENTRY *dll_callback)(HINSTANCE h, DWORD ul_reason_for_call, PVOID pv) = DllMain;
+static void (APIENTRY *dll_callback)(HINSTANCE h, DWORD ul_reason_for_call, PVOID pv) = OpenBLASDllMain;
 #ifdef _WIN64
 #pragma const_seg()
 #else

